### PR TITLE
py-pillow fixes

### LINKED ISSUE #14506: py-pillow fails test after building extension
The py-pillow package fails tests after building the extension. Here is the information from the build log:

```
=================================== FAILURES ===================================
_________________________ TestFilePalm.test_monochrome _________________________


self = <Tests.test_file_palm.TestFilePalm testMethod=test_monochrome>


    def test_monochrome(self):
        # Arrange
        mode = "1"
    
        # Act / Assert
        self.helper_save_as_palm(mode)
>       self.roundtrip(mode)


Tests/test_file_palm.py:38: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Tests/test_file_palm.py:30: in roundtrip
    self.assert_image_equal(converted, im)
Tests/helper.py:105: in assert_image_equal
    a.mode, b.mode, msg or "got mode {!r}, expected {!r}".format(a.mode, b.mode)
E   AssertionError: 'P' != '1'
E   - P
E   + 1
E    : got mode 'P', expected '1' 
```

### PR DESCRIPTION
**Pull request by aweits**
Fixes a couple of issues with `py-pillow`:

- [x] The unit tests were missing a dependency on `imagemagick` (see #14506)
- [x] The build was still using locally installed libraries for features that were disabled

Fixes #14506

### CODE REVIEW

**Comment by adamjstewart** [comment]
1. This should have been fixed by #14382, can you try again?
2. I think there's a cleaner way to do this, let me investigate.
3. Are you sure? On macOS, I'm seeing:
```console
$ otool -L $(find . -name '*.so') | sort | uniq
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/freetype-2.10.1-52goouhw7dni3yozivnw2py5d2x36yck/lib/libfreetype.6.dylib (compatibility version 24.0.0, current version 24.1.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/lcms-2.9-5duz5uysiwmclvgga3euiyinxa3lyfpg/lib/liblcms2.2.dylib (compatibility version 3.0.0, current version 3.8.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/libtiff-4.0.10-nlewmdvi5wqohlp3dp5waxe4djy3cl46/lib/libtiff.5.dylib (compatibility version 10.0.0, current version 10.0.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/libwebp-1.0.3-uj3feqesh44l7lpkkocstg4xe4zmi2pb/lib/libwebp.7.dylib (compatibility version 8.0.0, current version 8.5.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/libwebp-1.0.3-uj3feqesh44l7lpkkocstg4xe4zmi2pb/lib/libwebpdemux.2.dylib (compatibility version 3.0.0, current version 3.6.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/libwebp-1.0.3-uj3feqesh44l7lpkkocstg4xe4zmi2pb/lib/libwebpmux.3.dylib (compatibility version 4.0.0, current version 4.4.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/openjpeg-2.3.1-zp442otrfpxcdknldup7lq4p5nnc3oau/lib/libopenjp2.7.dylib (compatibility version 7.0.0, current version 2.3.1)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/zlib-1.2.11-cvrek7vtvob6v2dfnvdowu2rkqtqtmmi/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
	@rpath/libjpeg.62.dylib (compatibility version 62.0.0, current version 62.3.0)
```
so it appears to be linking to Spack's libraries correctly, although I may not have those same libraries available on my OS.

**Comment by aweits** [comment]
1.) Yes, it would seem that I missed seeing #14382. That is fixed.
2.) I'm not terribly enamored with that solution either :-)
3.) Yes, I'm sure. You can see in the build log that install re-calls build_ext, and then it suddenly finds all of the stuff I have installed locally.

```
coltrane (UPSTREAMSPACK/spack):spack find -Lv py-pillow
==> 1 installed package
-- linux-centos7-ivybridge / gcc@7.4.0 --------------------------
67uqks53jx4r6llzpzvsvop5anii2rya py-pillow@7.0.0~freetype~jpeg2000~lcms~tiff~webp~webpmux

coltrane (UPSTREAMSPACK/spack):find /home/aweits/GIT/UPSTREAMSPACK/spack/opt/spack/linux-centos7-ivybridge/gcc-7.4.0/py-pillow-7.0.0-67uqks53jx4r6llzpzvsvop5anii2rya -name \*.so -exec ldd {} \; | grep -v "spack" | awk '{ print $1 }' | sort | uniq
/lib64/ld-linux-x86-64.so.2
libcrypt.so.1
libc.so.6
libdl.so.2
libfreebl3.so
libfreetype.so.6
liblcms2.so.2
libm.so.6
libpng15.so.15
libpthread.so.0
libutil.so.1
linux-vdso.so.1
```

File: var/spack/repos/builtin/packages/py-pillow/package.py [hunk 1]
```diff
@@ -53,6 +53,7 @@ class PyPillow(PythonPackage):
     depends_on('jpeg')
     depends_on('py-pytest', type='test')
     depends_on('py-pytest-runner', type='test')
+    depends_on('imagemagick', type='test')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pillow/package.py]: I don't have imagemagick installed on my system, but the tests still pass for me. Is this dependency documented anywhere?

**Comment by adamjstewart** [comment]
Okay, #14670 should fix 2).

**Comment by adamjstewart** [comment]
With regards to 3), are you trying to build `py-pillow` with or without support for those libraries?

File: var/spack/repos/builtin/packages/py-pillow/package.py [hunk 2]
```diff
@@ -77,32 +78,48 @@ def patch(self):
...
+        def variant_to_cfg(setup):
+            able = 'enable' if '+' + variant in self.spec else 'disable'
+            return '{0}-{1}=1\n'.format(able, variant)
+
+        with open('setup.cfg', 'a') as setup:
+            # Default backend
+            setup.write('[build_ext]\n')
+            setup.write('enable-zlib=1\n')
+            setup.write('enable-jpeg=1\n')
+            setup.write('debug=1\n')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pillow/package.py]: This wasn't present before, why the change? Maybe this could be a variant.

File: var/spack/repos/builtin/packages/py-pillow/package.py (hunk 1 above)
Comment (aweits) [reply var/spack/repos/builtin/packages/py-pillow/package.py]: I do have imagemagick installed on the system but not in spack. See Tests/helper.py in Pillow.

File: var/spack/repos/builtin/packages/py-pillow/package.py (hunk 2 above)
Comment (aweits) [reply var/spack/repos/builtin/packages/py-pillow/package.py]: Oops, that should be removed.

**Comment by aweits** [comment]
So what remains is attempting to work around esoteric failures stemming from locally installed packages causing failures or getting compiled into spack packages. The real solution is build/test in some kind of container by default.

**Comment by aweits** [comment]
Ah, and yes trying to compile without support. The install phase finds the local packages and without the information provided on the command line, happily turns support for them on.

File: var/spack/repos/builtin/packages/py-pillow/package.py (hunk 1 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pillow/package.py]: Ah, I bet the issue is that since I don't have imagemagick, it skips those tests, but since you do have it, they are run and fail. Must be an issue with older versions of imagemagick. In that case, yes, we should add the dependency.

File: var/spack/repos/builtin/packages/py-pillow/package.py [hunk 3]
```diff
@@ -85,22 +86,26 @@ def patch(self):
...
-        args.extend(list(map(variant_to_flag, variants)))
-
-        # Spack does not (yet) support these modes of building
-        args.append('--disable-imagequant')
-
-        args.append('--rpath={0}'.format(':'.join(self.rpath)))
-        return args
+        self._pillow_setup()
+
+    def _pillow_setup(self):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pillow/package.py]: Let's just do this directly in `patch` instead of creating a new function.

Comment (aweits) [reply var/spack/repos/builtin/packages/py-pillow/package.py]: ok

File: var/spack/repos/builtin/packages/py-pillow/package.py [hunk 4]
```diff
@@ -86,21 +87,22 @@ def patch(self):
...
-        def variant_to_flag(variant):
-            able = 'enable' if '+' + variant in spec else 'disable'
-            return '--{0}-{1}'.format(able, variant)
-
-        args = ['--enable-zlib', '--enable-jpeg']
-
-        variants = ['tiff', 'freetype', 'lcms', 'webp', 'webpmux', 'jpeg2000']
-        args.extend(list(map(variant_to_flag, variants)))
-
-        # Spack does not (yet) support these modes of building
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pillow/package.py]: I would keep this comment below

**Comment by adamjstewart** [comment]
Okay, on macOS 10.15.2 with Python 3.7.6, I successfully built:

* `py-pillow@7.0.0+freetype+jpeg2000+lcms+tiff+webp+webpmux`:
```
--------------------------------------------------------------------
PIL SETUP SUMMARY
--------------------------------------------------------------------
version      Pillow 7.0.0
platform     darwin 3.7.6 (default, Jan 26 2020, 17:21:32)
             [Clang 11.0.0 (clang-1100.0.33.17)]
--------------------------------------------------------------------
--- JPEG support available
--- OPENJPEG (JPEG2000) support available (2.3)
--- ZLIB (PNG/ZIP) support available
*** LIBIMAGEQUANT support not available
--- LIBTIFF support available
--- FREETYPE2 support available
--- LITTLECMS2 support available
--- WEBP support available
--- WEBPMUX support available
--------------------------------------------------------------------
```
```console
$ otool -L $(find . -name '*.so') | sort | uniq
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/freetype-2.10.1-52goouhw7dni3yozivnw2py5d2x36yck/lib/libfreetype.6.dylib (compatibility version 24.0.0, current version 24.1.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/lcms-2.9-5duz5uysiwmclvgga3euiyinxa3lyfpg/lib/liblcms2.2.dylib (compatibility version 3.0.0, current version 3.8.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/libtiff-4.0.10-nlewmdvi5wqohlp3dp5waxe4djy3cl46/lib/libtiff.5.dylib (compatibility version 10.0.0, current version 10.0.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/libwebp-1.0.3-uj3feqesh44l7lpkkocstg4xe4zmi2pb/lib/libwebp.7.dylib (compatibility version 8.0.0, current version 8.5.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/libwebp-1.0.3-uj3feqesh44l7lpkkocstg4xe4zmi2pb/lib/libwebpdemux.2.dylib (compatibility version 3.0.0, current version 3.6.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/libwebp-1.0.3-uj3feqesh44l7lpkkocstg4xe4zmi2pb/lib/libwebpmux.3.dylib (compatibility version 4.0.0, current version 4.4.0)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/openjpeg-2.3.1-zp442otrfpxcdknldup7lq4p5nnc3oau/lib/libopenjp2.7.dylib (compatibility version 7.0.0, current version 2.3.1)
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/zlib-1.2.11-cvrek7vtvob6v2dfnvdowu2rkqtqtmmi/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
	@rpath/libjpeg.62.dylib (compatibility version 62.0.0, current version 62.3.0)
```
* `py-pillow@6.2.2~freetype~jpeg2000~lcms~tiff~webp~webpmux`:
```
--------------------------------------------------------------------
PIL SETUP SUMMARY
--------------------------------------------------------------------
version      Pillow 6.2.2
platform     darwin 3.7.6 (default, Jan 26 2020, 17:21:32)
             [Clang 11.0.0 (clang-1100.0.33.17)]
--------------------------------------------------------------------
--- JPEG support available
*** OPENJPEG (JPEG2000) support not available
--- ZLIB (PNG/ZIP) support available
*** LIBIMAGEQUANT support not available
*** LIBTIFF support not available
*** FREETYPE2 support not available
*** LITTLECMS2 support not available
*** WEBP support not available
*** WEBPMUX support not available
--------------------------------------------------------------------
```
```console
$ otool -L $(find . -name '*.so') | sort | uniq
	/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/zlib-1.2.11-cvrek7vtvob6v2dfnvdowu2rkqtqtmmi/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
	@rpath/libjpeg.62.dylib (compatibility version 62.0.0, current version 62.3.0)
```
So this change is safe to merge as far as I'm concerned.
